### PR TITLE
CASSANDRA-14204 trunk

### PR DESCRIPTION
**Pull request by smiklosovic**
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

### CODE REVIEW

File: src/java/org/apache/cassandra/db/compaction/CompactionManager.java [hunk 1]
```diff
@@ -680,12 +680,28 @@ public AllSSTableOpStatus performGarbageCollection(final ColumnFamilyStore cfSto
...
+                        SSTableReader sstable = iter.next();
+                        if (!sstable.isRepaired())
+                        {
+                            try
+                            {
+                                transaction.cancel(sstable);
+                            }
+                            finally
+                            {
+                                iter.remove();
```
Comment (blambov) [inline src/java/org/apache/cassandra/db/compaction/CompactionManager.java]: Removal in the interior of the array list makes this an O(n^2) operation, which can be a problem when the number of sstables can grow proportionally with data size (e.g. with LCS and UCS).
I'd turn this into an iteration over the source, deciding to add to an initially empty array or cancel.

**Review by blambov — COMMENTED** [review]
The fix makes sense.

We may want to consider not removing unrepaired sstables from the list at all. Garbagecollect removes deleted data, and like any compaction it makes sure to keep any tombstone that is not repaired. It's likely that it does not help as much with unrepaired sstables as it does with repaired ones, but it makes it possible to remove tombstones immediately after an sstable attains repaired status.

**Comment by smiklosovic** [comment]
@blambov would  you mind to take a look again, please?

File: src/java/org/apache/cassandra/db/compaction/CompactionManager.java [hunk 2]
```diff
@@ -680,12 +680,25 @@ public AllSSTableOpStatus performGarbageCollection(final ColumnFamilyStore cfSto
...
             public Iterable<SSTableReader> filterSSTables(LifecycleTransaction transaction)
             {
-                Iterable<SSTableReader> originals = transaction.originals();
+                List<SSTableReader> filteredSSTables = new ArrayList<>();
                 if (cfStore.getCompactionStrategyManager().onlyPurgeRepairedTombstones())
-                    originals = Iterables.filter(originals, SSTableReader::isRepaired);
-                List<SSTableReader> sortedSSTables = Lists.newArrayList(originals);
-                Collections.sort(sortedSSTables, SSTableReader.maxTimestampAscending);
-                return sortedSSTables;
+                    for (SSTableReader sstable : transaction.originals())
```
Comment (blambov) [inline src/java/org/apache/cassandra/db/compaction/CompactionManager.java:686]: Nit: Multi-line cases are usually enclosed in {}. In this case there's some potential for confusion about the else case on 697, thus at least the for cycle needs to be in a block.

**Review by blambov — APPROVED** [review]

**Pull request closed without merging**